### PR TITLE
Add Mutation Engine to API

### PR DESCRIPTION
**Pull request by RlonRyan**
The mutation engine can now be customized through the AgriCraft API.

### CODE REVIEW

File: src/main/java/com/infinityraider/agricraft/AgriCraft.java [hunk 1]
```diff
@@ -108,18 +108,21 @@ public void registerMessages(INetworkWrapper wrapper) {
 	@SuppressWarnings("unused")
 	public void onPreInit(FMLPreInitializationEvent event) {
 		super.preInit(event);
+        PluginHandler.preInit(event);
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/AgriCraft.java]: This has been moved to the proxy
https://github.com/InfinityRaider/AgriCraft/blob/master/src/main/java/com/infinityraider/agricraft/proxy/IProxy.java#L30

Comment (RlonRyan) [inline src/main/java/com/infinityraider/agricraft/AgriCraft.java]: Oh, so that's where that went...

**Comment by RlonRyan** [comment]
So... is this a go?

File: src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java [hunk 2]
```diff
@@ -21,12 +21,22 @@
 import javax.annotation.Nonnull;
 import java.util.*;
 
+/**
+ * Class representing possible custom wood types.
+ * 
+ * This class is candidate for a rewrite/cleaning.
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java]: why?

File: src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java [hunk 3]
```diff
@@ -90,27 +99,19 @@ public NBTTagCompound writeToNBT(NBTTagCompound tag) {
...
-            return getDefault();
-        }
-        if (!tag.hasKey(AgriNBT.MATERIAL)) {
-            AgriCore.getLogger("AgriCraft").debug("TECW: Tag missing material!");
-            return getDefault();
-        }
-        if (!tag.hasKey(AgriNBT.MATERIAL_META)) {
-            AgriCore.getLogger("AgriCraft").debug("TECW: Tag missing meta!");
+        // The old code was much too annoying, in terms of log spam.
+        if (!NBTHelper.hasKey(tag, AgriNBT.MATERIAL_META, AgriNBT.MATERIAL)) {
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java]: I like that method

File: src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java (hunk 2 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java]: Have you seen it? It's quite messy.

Comment (RlonRyan) [inline src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java]: That being said, it does work...

File: src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java [hunk 4]
```diff
@@ -90,27 +99,19 @@ public NBTTagCompound writeToNBT(NBTTagCompound tag) {
...
-            AgriCore.getLogger("AgriCraft").debug("TECW: Passed Null Tag!");
-            return getDefault();
-        }
-        if (!tag.hasKey(AgriNBT.MATERIAL)) {
-            AgriCore.getLogger("AgriCraft").debug("TECW: Tag missing material!");
-            return getDefault();
-        }
-        if (!tag.hasKey(AgriNBT.MATERIAL_META)) {
-            AgriCore.getLogger("AgriCraft").debug("TECW: Tag missing meta!");
+        // The old code was much too annoying, in terms of log spam.
```
Comment (RlonRyan) [inline src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java]: It doesn't like jei, which causes it to spew "invalid tag" into the log.

File: src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java [hunk 5]
```diff
@@ -90,27 +99,19 @@ public NBTTagCompound writeToNBT(NBTTagCompound tag) {
...
     public boolean equals(Object obj) {
-        if(obj == this) {
+        if (obj == this) {
             return true;
-        }
-        if(!(obj instanceof CustomWoodType)) {
+        } else if (obj instanceof CustomWoodType) {
+            CustomWoodType other = (CustomWoodType) obj;
+            return other.getBlock() == this.getBlock() && other.getMeta() == this.getMeta();
+        } else {
```
Comment (RlonRyan) [inline src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java]: Fewer lines of code.

**Comment by RlonRyan** [comment]
Can we merge this please and start a formatting branch?

**Comment by RlonRyan** [comment]
I say that because I feel we are getting off track, and this pr contains important code, which is getting hidden by the formatting changes...

File: src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java (hunk 3 above)
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/utility/CustomWoodType.java]: NBTHelper.hasKey(), no more if(hasKey() && hasKey() && ...

**Pull request merged**